lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs

```diff
@@ -56,27 +51,30 @@ internal sealed class TaskLifeCycle
 
         public void Start() 
         {
-            if (!_taskStart.IsPresent())
+            if (Interlocked.Exchange(ref _startHasBeenInvoked, 1) == 0)
             {
-                return;
-            }
-
-            foreach (var startHandler in _taskStartHandlers)
-            {
-                startHandler.OnNext(_taskStart.Value);
+                foreach (var startHandler in _taskStartHandlers)
+                {
+                    startHandler.OnNext(_taskStart);
+                }
             }
         }
 
         public void Stop() 
         {
-            if (!_taskStop.IsPresent())
+            try
             {
-                return;
+                if (Interlocked.Exchange(ref _stopHasBeenInvoked, 1) == 0)
+                {
+                    foreach (var stopHandler in _taskStopHandlers)
+                    {
+                        stopHandler.OnNext(_taskStop);
+                    }
+                }
             }
-
-            foreach (var stopHandler in _taskStopHandlers)
+            catch (Exception e)
             {
-                stopHandler.OnNext(_taskStop.Value);
+                throw new TaskStopHandlerException("Encountered Exception on TaskStopHandler.", e);
```

Comment:
Do we serialize TaskStopHandlerException to evalautor or eventually replaced it with SystemException in ContextRuntime? This is actually the exception from client's handler, right?